runtimes/pythonrt/ak_runner.py

```diff
@@ -0,0 +1,343 @@
+"""Run user code under AutoKitteh"""
+
+# This file is long, but keeping it a single file helps with embedding it in Go and
+# running it.
+
+import ast
+import builtins
+import json
+import logging
+import pickle
+import sys
+import tarfile
+from base64 import b64decode, b64encode
+from functools import wraps
+from importlib.abc import Loader
+from importlib.machinery import SourceFileLoader
+from inspect import isbuiltin
+from os import mkdir
+from pathlib import Path
+from queue import Queue
+from socket import AF_UNIX, SOCK_STREAM, socket
+from threading import Thread
+
+# TODO: Log to AutoKitteh (ENG-552)
+logging.basicConfig(
+    format='%(asctime)s - %(levelname)s - %(filename)s:%(lineno)d - %(message)s',
+    datefmt='%Y-%M-%DT%H:%M:%S',
+    level=logging.INFO,
+)
+
+
+def name_of(node):
+    """Name of call node (e.g. 'requests.get')"""
+    if isinstance(node, ast.Name):
+        return node.id
+
+    prefix = name_of(node.value)
+    return f'{prefix}.{node.attr}'
+
+
+ACTION_NAME = '_ak_call'
+MODULE_NAME = ''
+BUILTIN = {v for v in dir(builtins) if callable(getattr(builtins, v))}
+
+
+class Transformer(ast.NodeTransformer):
+    """Replace 'fn(a, b)' with '_ak_call(fn, a, b)'"""
+    def visit_Call(self, node):
+        name = name_of(node.func)
+        # ast.Transformer does not recurse to args
+        node.args = [self.visit(a) for a in node.args]
+
+        if not name or name in BUILTIN:
+            return node
+
+        logging.info('patching %s with action', name)
+        call = ast.Call(
+            func=ast.Name(id=ACTION_NAME, ctx=ast.Load()),
+            args=[node.func] + node.args,
+            keywords=node.keywords,
+        )
+        return call
+
+
+class AKLoader(Loader):
```

Comment:
I think i'm missing something here. Why do we need this? Don't we just capture all the imports and wrap them? Why do we need to deal with the AST?